app/src/main/java/com/example/booklisting/MainActivity.java

```diff
@@ -130,38 +125,42 @@ public void onLoadMore() {
     }
 
     @Override
-    public void onLoadFinished(@NonNull Loader<List<Book>> loader, List<Book> list) {
+    public void onLoadFinished(@NonNull Loader<BookLoaderResult> loader, BookLoaderResult loaderResult) {
         if ((adapter == null) || (adapter.getItemCount() == 0)) {
-            if (exceptionThrown) {
+            if (loaderResult.getException() != null) {
+                Log.i("Exception is", String.valueOf(loaderResult.getException()));
                 progressBar.setVisibility(View.GONE);
                 emptyView.setVisibility(View.GONE);
                 showAlert();
             } else {
 
-                if (list.isEmpty()) {
+                if (loaderResult.getResult().isEmpty()) {
                     progressBar.setVisibility(View.GONE);
                     emptyView.setVisibility(View.VISIBLE);
                     emptyView.setText(R.string.no_books_found);
                 } else {
                     emptyView.setVisibility(View.GONE);
                     progressBar.setVisibility(View.GONE);
-                    books.addAll(list);
+                    books.addAll(loaderResult.getResult());
                     adapter = new BookAdapter(this, R.layout.list_item, (ArrayList<Book>) books);
                     adapter.setStateRestorationPolicy(RecyclerView.Adapter.StateRestorationPolicy.PREVENT_WHEN_EMPTY);
                     bookList.setAdapter(adapter);
                 }
             }
         } else {
-            if (exceptionThrown) {
+            if (loaderResult.getResult().isEmpty()) {
                 emptyView.setVisibility(View.GONE);
                 progressBar.setVisibility(View.GONE);
-                showAlert();
+                if (books.get(books.size() - 1) == null) {
```

Comment:
I'm not clear what is this check actually checking?

Reply:
It was naive solution for problem that i had with deleting last entry, when new set of data is loaded. I think that that bug was resolved whit proper implementation of onScrolllistener. Anyway it works fine now, so I removed that check.